File: leetcode/429-n-ary-tree-level-order-traversal/main.go
package main

type NaryNode struct {
	Val      int
	Children []*NaryNode
}

// bfs, iterative
func levelOrder(root *NaryNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	var result [][]int
	queue := []*NaryNode{root}
	for len(queue) > 0 {
		lengthOnSameLevel := len(queue)
		nodesOnSameLevel := []int{}
		for i := 0; i < lengthOnSameLevel; i++ {
			head := queue[0]
			queue = queue[1:len(queue)]
			nodesOnSameLevel = append(nodesOnSameLevel, head.Val)
			for j := 0; j < len(head.Children); j++ {
				if head.Children[j] != nil {
					queue = append(queue, head.Children[j])
				}
			}
		}
		result = append(result, nodesOnSameLevel)
	}
	return result
}

func main() {

}

File: leetcode/429-n-ary-tree-level-order-traversal/main.py
class Node(object):
    def __init__(self, val, children):
        self.val = val
        self.children = children


"""
    1st: BFS
    - level order traversal

    Time    O(N)
    Space   O(N)
    52 ms, faster than 81.93%
"""


class Solution(object):
    def levelOrder(self, root):
        if root == None:
            return []
        res = []
        q = [root]
        while len(q) > 0:
            n = len(q)
            arr = []
            for _ in range(n):
                node = q.pop(0)
                arr.append(node.val)
                for child in node.children:
                    q.append(child)
            res.append(arr)
        return res


#       1
#   2   3   4
#  5 6
# 7
a = Node(1, [])
b = Node(2, [])
c = Node(3, [])
d = Node(4, [])
e = Node(5, [])
f = Node(6, [])
g = Node(7, [])

a.children = [b, c, d]
b.children = [e, f]
e.children = [g]

ans = Solution().levelOrder(a)
print("ans=", ans)
